atef/config.py

```diff
@@ -1207,6 +1207,21 @@ async def _compare(self, data: Any) -> Result:
             identifier=self.identifier
         )
 
+    async def compare(self) -> Result:
+        """
+        Run the inherited compare routine, but if signal is an enum,
+        also flip the string field and retry
+        """
+        result = await super().compare()
+
+        if (result.severity != Severity.success
+                and getattr(self.signal, 'enum_strs', None)):
+            self.comparison.string = not self.comparison.string
```

Comment:
This sets it and never resets it? That seems not good

It would seem like to me that we could be a bit smarter and get the string value from the enum strings list ourselves and redo the comparison
"If the comparison failed and we have an integer from the control layer, get the corresponding string value and run a new simple comparison. If and only if that succeeds, adjust the `result` (otherwise stick with the original failure)"

Reply:
We should definitely reset the value, but I wanted to keep the data-gathering logic out of this class to avoid fragmenting that logic across multiple locations.

Reply:
You're already digging into the signal to pull out enum strings in the comparison - that seems to be reaching beyond where it should.

This specific PR is trying to figure out the user's intent ("oh they must not know what they're asking for, let's try the other method"). The original idea was to make the user specify whether they wanted a string comparison or an integer comparison by selecting "string". What happens if there's some overlap/ambiguity in the values? e.g., enum strings `["0", "2", "foobar"]`, and the comparison wants to know if the value is "2"? I suppose either "2" or "foobar" would be valid in that case, rather strangely.

If that's still not a concern, perhaps it'd be worthwhile to consider a special wrapper for enum values at the signal caching / value retrieval layer. That is to say a wrapper that says `1 == "Enum Value for 1"` is `True`

Reply:
To me there was a distinction between checking if enum_strs exists on the signal (which is already stored on the `PreparedSignalComparison`) and using the cache to get the signal value (`self.cache.get_signal_data`).  Are those two less distinct than I thought?  

----------------------------------

The `["0", "2", "foobar"]` case is interesting, and I hope nobody makes an enum like this...  A bit of me detailing my thought process below.
If the user wants to check the signal against `"2"` (string), getting the data from the signal will only return one of the enum values.  
* if signal has "2" (index 1), the signal could either return `"2"` (string) or `1` (int)
  * comparison passes on the first case
* if signal has "foobar" (index 2), the signal could either return `"foobar"` (string) or `2` (int)
  * comparison fails on both cases 

The key here is that we're only ever comparing to an enum in one of the positions.

If we do try to grab the enum value corresponding to the user provided "2", then the potential for confusion arises.  The comparison's "string" setting only changes what we get from the signal, not the user's provided value.  

I think the upshot of the above block of text is I feel like the current process isn't too ambiguous.

----------

Having finally made it to the last bit of your comment I like that idea better than all the above 🤣